.github/action_scripts/delegated_staking/lib.js

```diff
@@ -0,0 +1,355 @@
+  })
+
+  const expectedPendingLength = pendingStakes.length
+  const actualPendingLength = stakeResponse.pendingWithdrawals.length
+  if (expectedPendingLength !== actualPendingLength) {
+    throw new Error(
+      `Expected ${expectedPendingLength} active stakes but got ${actualPendingLength}`,
+    )
+  }
+
+  Object.values(pendingStakes).map((stakeItem) => {
+    assertAllKeysMatch(stakeResponse.pendingWithdrawals, stakeItem)
+  })
+}
+
+// Get stake to update and wait until it has some rewards
+const fetchStakeWithRewardsBalance = async (
+  urls,
+  address,
+  stakeHash,
+  nodeId = null,
+) => {
+  return withRetry(
+    async () => {
+      const stakeResponse = await getAccountDelegatedStakes(urls, address)
+      const stake = stakeResponse.activeDelegatedStakes.find(
+        (stake) => stake.hash === stakeHash && stake.rewardAmount > 0,
+      )
+
+      if (!stake) {
+        throw new Error('Stake not found with rewards balance')
+      }
+
+      const stakeAlreadyExists = stakeResponse.activeDelegatedStakes.find(
+        (stake) => {
+          return (
+            (nodeId ? stake.nodeId === nodeId : true) &&
+            address === stake.source
+          )
+        },
+      )
+
+      if (stakeAlreadyExists) {
+        throw new Error('Cant update, stake already exists')
+      }
+
+      return stake
+    },
+    {
+      name: 'FetchStakeWithRewardsBalance',
+      maxAttempts: 20,
+      interval: 5 * 1000,
+      handleError: () => {},
+    },
+  )
+}
+
+const createTokenLock = async (account, urls, lockAmount) => {
+  const initialBalance = dagToDatum(await account.getBalance())
+
+  const { hash } = await account.postTokenLock({
+    source: account.address,
+    amount: lockAmount,
+    tokenL1Url: urls.dagL1Url,
+    unlockEpoch: null,
+    currencyId: null,
+    fee: 0,
+  })
+
+  if (!hash) {
+    throw new Error('Failed to create TokenLock')
+  }
+
+  await withRetry(
+    async () => assertBalanceChange(account, initialBalance - lockAmount),
+    {
+      name: 'assertBalanceChangeAfterTokenLock',
+      maxAttempts: 10,
+      interval: 1000,
+      handleError: () => {},
+    },
+  )
+
+  return hash
+}
+
+const assertBalanceChange = async (account, expectedBalanceDatum) => {
+  const balance = dagToDatum(await account.getBalance())
+
+  if (balance !== expectedBalanceDatum) {
+    throw new Error(
+      `Invalid balance: Expected balance to be ${expectedBalanceDatum} but got ${balance}`,
+    )
+  }
+}
+
+const getNodeParams = async (urls) => {
+  logWorkflow.info(`Request to: ${urls.globalL0Url}/node-params`)
+  const response = await axios.get(
+    `${urls.globalL0Url}/node-params?t=${Date.now()}`,
+    {
+      headers: {
+        'Cache-Control': 'no-cache, no-store, must-revalidate',
+        Pragma: 'no-cache',
+        Expires: '0',
+      },
+    },
+  )
+  checkOk(response)
+  return response.data
+}
+
+const fetchSnapshot = async (urls, ordinal) => {
```

Comment:
I think a utility like this that doesn't pertain to delegated staking doesn't belong in this file